### PR TITLE
Exclude directories from beeing processed

### PR DESCRIPTION
**Pull request by mmattel**
Please see also #15478 (Functionality: exlcude directories from beeing processed) which I needed to close due to a git problem I was not able to solve.

Summary:
1. Enterprise Storage Systems are capable of Snapshots. These Snapshots are directories with a particular name and keep point in time views of the data. Usually, these snapshot directories are present in each directory
2. There is no common naming for these directories and will never be. NetApp uses `.snapshot` and `~snapshot`, EMC eg `.ckpt`, HDS eg `.latest` and `~latest` and so on
3. Viewing and scanning of these directories does not make any sense as these directories are used to restore files from an OS point of view
4. You can also have hardlinks to backups with the same background as described in #15385
5. Scanning a huge datastore containing snapshots does mean that for each snapshot taken ALL files of the datastore are imported into the database again. Example: a datastore containing 200K files, having 15 snapshots means scanning of 3M files where 2.8M are worthless to be imported

There are for sure more resons, but I tried to keep it short.

With this PR, you can define a array of directory names in config.php which are further excluded from processing. It extends the current functionality of excluding `.` and `..`

@DeepDiver1975 @karlitschek @nickvergessen @Aesculapius

### CODE REVIEW

**Comment by karlitschek** [comment]
@LukasReschke @DeepDiver1975 @icewind1991

File: lib/private/files/filesystem.php [hunk 1]
```diff
@@ -584,6 +585,14 @@ static public function isIgnoredDir($dir) {
...
 		}
+		$excluded = \OC::$server->getSystemConfig()->getValue('excluded_directories', array());
+		if (!empty($excluded)) {
+			foreach ($excluded as $ex) {
+				if ($ex === $dir) {
+					return true;
+				}
+			}
+		}
 		return false;
```
Comment (nickvergessen) [inline lib/private/files/filesystem.php]: Why don't you use `return in_array($dir, $excluded);` like we do for the files option?

**Comment by LukasReschke** [comment]
I'm still not absolutely sure whether this is the best way to implement this, for one we need to ensure that all code paths call `isIgnoredDir` so that you can't upload files there as such invisible files would be very confusing for users. cc @icewind1991

**Comment by LukasReschke** [comment]
(also requires unit tests)

**Comment by mmattel** [comment]
It is impotant that we talk about directories and not files. What means that if a directory is not "visible" via owncloud, you can not write into.

In the first step, it is a matter of scanning. What is not scanned is not shown and further not processed.

But you are right about the ``creation of a directory via owncloud`` with an excluded directory name, that should be covered.

In filesystem.php, we have the function:

	static public function mkdir($path) {
		return self::$defaultInstance->mkdir($path);
	}

What about if we do:

	static public function mkdir($path) {
		if (self::isIgnoredDir(basename($path))) {
			return false;
		}
		return self::$defaultInstance->mkdir($path);
	}

I have tested this and this works. I get a owncloud error in case a excluded directory is chosen:
image

If I comment out the code of ``mkdir`` I added and try to create a directory called ``~snapshot`` into a SMB share which is on a storage having snapshot directory names like ``~snapshot`` I get: 
image
This means, that the code above would perfectly work. Because anyway, the directory can not be created via the owncloud - it returns an error.
Please let me know your thoughts.

But it also would be great to cover that in an earler stage eg when the path is entered: https://github.com/owncloud/core/blob/master/apps/files/js/files.js#L97
Here I do not know how to do it, I would need some help.

**Comment by nickvergessen** [comment]
@mmattel but we have several ways of doing that, webdav, provisioning api, web ui, ...
and that is what lukas means, it's hard to make sure that all use this way now all of a sudden.

**Comment by mmattel** [comment]
OK, I got that. Step by step.

I added a check at the webgui for excluded directories and that works. If you try to create a directory via the webgui with a name in the excluded list, you get a message presented.

Where can I look next to adopt the code for implementing this?

**Comment by mmattel** [comment]
Changes to cover webdav, webgui and client access

Following test have been made creating either a file or directory having once a allowed and once a excluded name.

The functions shown in the table are except one all triggered  and covered by the access methods used in lib/private/files/view.php

For WebDav, CyberDuck was used on a W7 client.
The IOS client run the latest available version (3.4.1)
All tests made passed well.
A unit test is open to be made.

		| webdav	| webgui	| client IOS
------------ |------------ |------------ | -------------
directory create | mkdir	| mkdir		| mkdir
directory rename | rename	| rename	| rename
directory delete | rmdir	| rmdir		| rmdir
		 | 		| 		| 
file create	 | put(*)	| touch		| n.a.
file rename	 | rename	| rename	| rename
directory delete | unlink	| unlink	| unlink


n.a. ... function not available
(*) put ... this function is not handled in lib/private/files/view.php  but in lib/private/connector/sabre/directory.php which points to /sabre/file.php

**Comment by mmattel** [comment]
git drives me nuts, it was just a rebase and now I have all those additional commits in...
will try to fix that.

**Comment by mmattel** [comment]
Closing due to a rebase problem will reopen with a new PR.
Sorry for any inconvenience caused

**Pull request closed without merging**